source/common/conn_pool/conn_pool_base.cc

```diff
@@ -571,5 +572,20 @@ void ActiveClient::onConnectTimeout() {
   close();
 }
 
+void ActiveClient::drain() {
+  if (currentUnusedCapacity() <= 0) {
+    return;
+  }
```

Comment:
How can this be below zero?

Reply:
If we start with a stream limit of 10, have 10 outstanding streams, and get a SETTINGs frame decreasing streams/connection to 5.  We'd theoretically pretty immediately get resets for the 5 extra streams but it can go negative during the interim.